### PR TITLE
pilot-agent add cert expiry metric

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

add a Gauge metric of `default` cert expiry minutes on `pilot-agent`

```console
# HELP istio_agent_cert_expiry_minutes The left minutes, when cert chain will expire. A negative value indicates the cert is expired.
# TYPE istio_agent_cert_expiry_minutes gauge
istio_agent_cert_expiry_minutes{resource_name="default"} 701
```

fixed: https://github.com/istio/istio/issues/37291

***warning***

see from 
https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/metric/registry.go#L237

the registry use metric.Name as key, I think we can not have multi DerivedGauge (same name but different in labels) before we modify the implement of upstream.

In short-term, this PR will allow user monitor `default` cert.
for a long-term, we should make `DerivedGauge ` support metrics with different labels.

BTW, `opencensus-go` is merged to `OpenTelemetry`, so that would take a little more time to achieve this.

### CODE REVIEW

**Comment by anannaya** [comment]
Thanks for the PR .much needed one

File: security/pkg/nodeagent/cache/monitoring.go [hunk 1]
```diff
@@ -44,6 +51,12 @@ var (
 	numFileSecretFailures = monitoring.NewSum(
 		"num_file_secret_failures_total",
 		"Number of times secret generation failed for files")
+
+	certExpiryMinutes = monitoring.NewGauge(
+		"cert_expiry_minutes",
```
Comment (howardjohn) [inline security/pkg/nodeagent/cache/monitoring.go:56]: seconds may be more canonical: https://prometheus.io/docs/practices/naming/#base-units

File: security/pkg/nodeagent/cache/secretcache.go [hunk 2]
```diff
@@ -186,11 +186,23 @@ func NewSecretManagerClient(caClient security.Client, options *security.Options)
...
 	return ret, nil
 }
 
+func (sc *SecretManagerClient) refreshMetrics() {
+	t := time.NewTicker(time.Minute)
+	for range t.C {
+		if sc.cache.GetWorkload() != nil {
+			item := sc.cache.GetWorkload()
+			delay := sc.rotateTime(*item)
+			RecordCertExpiry(item.ResourceName, delay)
```
Comment (howardjohn) [inline security/pkg/nodeagent/cache/secretcache.go:201]: I think we want NewDerivedGauge, so we do not need to manually run a goroutine. Can compute it during each prometheus fetch

Comment (zirain) [reply security/pkg/nodeagent/cache/secretcache.go:201]: https://github.com/istio/pkg/blob/4fff90a48d5b5b21f7ab31a18db3b94272249b54/monitoring/monitoring.go#L189

why `NewDerivedGauge` not support labels? for now, I can just put resource name on metric name, and it will looks like as following

```console
# HELP istio_agent_default_cert_expiry_seconds The left time, in seconds, when cert chain will expire. A negative value indicates the cert is expired.
# TYPE istio_agent_default_cert_expiry_seconds gauge
istio_agent_default_cert_expiry_seconds 86242.038222406

```

the format looks wired

File: security/pkg/nodeagent/cache/monitoring.go [hunk 3]
```diff
@@ -55,3 +60,13 @@ func init() {
 		numFileSecretFailures,
 	)
 }
+
+func RecordCertExpiry(resourceName string, expiry time.Time) {
+	monitoring.NewDerivedGauge(
```
Comment (bianpengyuan) [inline security/pkg/nodeagent/cache/monitoring.go:65]: Can we add labels to derived metric? I know it is currently supported but I don't think there should be anything blocking?

Comment (bianpengyuan) [reply security/pkg/nodeagent/cache/monitoring.go:65]: Feels wrong to have resource name at stats name.

Comment (zirain) [reply security/pkg/nodeagent/cache/monitoring.go:65]: need to take a look at istio/pkg package

Comment (zirain) [reply security/pkg/nodeagent/cache/monitoring.go:65]: looks good, need to update `istio/pkg` first.

```console
# HELP istio_agent_cert_expiry_seconds The left time, in seconds, when cert chain will expire. A negative value indicates the cert is expired.
# TYPE istio_agent_cert_expiry_seconds gauge
istio_agent_cert_expiry_seconds{resource_name="default"} 86108.240811166
```

**Comment by zirain** [comment]
https://github.com/istio/pkg/pull/483 @howardjohn PTAL

**Comment by zirain** [comment]
here is the latest output:
```console
# HELP istio_agent_cert_expiry_seconds The left time, in seconds, when cert chain will expire. A negative value indicates the cert is expired.
# TYPE istio_agent_cert_expiry_seconds gauge
istio_agent_cert_expiry_seconds{resource_name="default"} 86279.188321444

```

**Review by bianpengyuan — APPROVED** [review]

File: security/pkg/nodeagent/cache/monitoring.go [hunk 4]
```diff
@@ -55,3 +59,17 @@ func init() {
 		numFileSecretFailures,
 	)
 }
+
+func recordCertExpiry(resourceName string, expiry time.Time) {
+	labels := map[string]string{
+		"resource_name": resourceName,
+	}
+	monitoring.NewDerivedGaugeWithLabels(
```
Comment (douglas-reid) [inline security/pkg/nodeagent/cache/monitoring.go:67]: what happens if this is called more than once with different resourceNames?

Comment (zirain) [reply security/pkg/nodeagent/cache/monitoring.go:67]: the metric will be updated.
```golang
	recordCertExpiry(item.ResourceName, item.ExpireTime)
	recordCertExpiry(item.ResourceName+"bak", item.ExpireTime)
```
output:
```console
# HELP istio_agent_cert_expiry_seconds The left time, in seconds, when cert chain will expire. A negative value indicates the cert is expired.
# TYPE istio_agent_cert_expiry_seconds gauge
istio_agent_cert_expiry_seconds{resource_name="defaultbak"} 86355.853291707

```

Comment (douglas-reid) [reply security/pkg/nodeagent/cache/monitoring.go:67]: Yes. I guess I'm asking if there is ever a valid case for two cert expiry metrics in the same process.

Comment (zirain) [reply security/pkg/nodeagent/cache/monitoring.go:67]: IIRC, only `default` cert, @howardjohn can you confirm this?

Comment (howardjohn) [reply security/pkg/nodeagent/cache/monitoring.go:67]: We support file mounted ones, so may be unbounded number actually. Also ROOTCA may be something we want to measure as well..?

Comment (douglas-reid) [reply security/pkg/nodeagent/cache/monitoring.go:67]: I think the workaround we want is something like: https://github.com/istio/pkg/pull/483/files#r810176904

Comment (zirain) [reply security/pkg/nodeagent/cache/monitoring.go:67]: https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/metric/registry.go#L237

the `registry` use `metric.Name` as key, I think we can not have multi `DerivedGauge` (same name but different in labels) before we modify the implement of upstream.

**Comment by costinm** [comment]
I assume the main goal here is to generate an alert if something is wrong - certs not renewed and close to expiration, and critical for expired certs.

For this purpose, time in seconds may not be the best metric.

We know when we are supposed to renew, so adding metric on how many certs are beyond the expected renewal time, and one for expired certs may be more useful. Or number of renewal failure. 

It is a critical area, we should also have some best practice, example alerts and in general a higher level view on this.

Not arguing against this metric, we can also expose this - but it's harder to page or know something is about to be wrong without some guessing.

**Comment by Christoph-Raab** [comment]
> I assume the main goal here is to generate an alert if something is wrong - certs not renewed and close to expiration, and critical for expired certs.

As requester of the feature I want to clarify: Our rootca certificate lifetime is rather short, speaking of days (not the default 10 years). We want to be able to monitor the workload certificates expiry to ensure that they do not expire after the root/intermediate cacerts, which would make it impossible for the sidecar to renew the workload certificate.

For that use case the time in seconds until expiry is exactly what we're looking for.

**Comment by zirain** [comment]
@howardjohn @douglas-reid updated the desc, please review this PR again.

File: security/pkg/nodeagent/cache/monitoring.go (hunk 4 above)
Comment (howardjohn) [reply security/pkg/nodeagent/cache/monitoring.go:67]: What is current status? Last one wins?

If so that is not a great option IMO. Either we only do `default`, do `default`+`ROOTCA`, only do the oldest, or do all of them. But doing the last one will give unexpected results

Comment (douglas-reid) [reply security/pkg/nodeagent/cache/monitoring.go:67]: @zirain, I was OOO so I may have lost context, but the metric.Name as the key shouldn't be an issue. If the function is applied with labels, it should be possible to have multiple values. I'll try to add the functionality with tests and ping back here.

Comment (zirain) [reply security/pkg/nodeagent/cache/monitoring.go:67]: > What is current status? Last one wins?
> 
> If so that is not a great option IMO. Either we only do `default`, do `default`+`ROOTCA`, only do the oldest, or do all of them. But doing the last one will give unexpected results

the last win, but right now it will only be called at `registerSecret`.

Comment (zirain) [reply security/pkg/nodeagent/cache/monitoring.go:67]: > @zirain, I was OOO so I may have lost context, but the metric.Name as the key shouldn't be an issue. If the function is applied with labels, it should be possible to have multiple values. I'll try to add the functionality with tests and ping back here.

if we can make `DerivedGauge` support metrics with same name and labels keys but different label values, everything will be more easier.

Comment (douglas-reid) [reply security/pkg/nodeagent/cache/monitoring.go:67]: @zirain @howardjohn please take a look at https://github.com/istio/pkg/pull/490. i think that (or similar) should allow us to define the metric once and have multiple calls like:

```golang
certExpirySeconds = monitoring.NewDerivedGauge(
       "cert_expiry_seconds",
       "The left time, in seconds, when cert chain will expire. ",
       monitoring.WithLabelKeys("resource_name"),
)
...
certExpirySeconds.ValueFrom(func() float64 { return time.Until(expiry).Seconds() }, resourceName)
```

File: go.mod [hunk 5]
```diff
@@ -2,6 +2,8 @@ module istio.io/istio
 
 go 1.17
 
+replace istio.io/pkg => github.com/douglas-reid/pkg v0.0.0-20220301021643-00c7529ce85a
```
Comment (howardjohn) [inline go.mod:5]: can remove now I think ?

**Review by howardjohn — APPROVED** [review]

File: go.mod (hunk 5 above)
Comment (zirain) [reply go.mod:5]: yes, that's a nit.

File: security/pkg/nodeagent/cache/monitoring.go [hunk 6]
```diff
@@ -44,6 +46,12 @@ var (
 	numFileSecretFailures = monitoring.NewSum(
 		"num_file_secret_failures_total",
 		"Number of times secret generation failed for files")
+
+	certExpirySeconds = monitoring.NewDerivedGauge(
+		"cert_expiry_seconds",
+		"The left time, in seconds, when cert chain will expire. "+
```
Comment (douglas-reid) [inline security/pkg/nodeagent/cache/monitoring.go:52]:
```suggestion
		"The time remaining, in seconds, before the certificate chain will expire. "+
```

**Comment by nrjpoddar** [comment]
Is this enabled by default?

**Comment by howardjohn** [comment]
Yes
